checker: discriminated-union narrowing leaks through nested property accesses (false-negative TS2339/TS2322)

**AgentName:** M1-Opus

## Summary

tsz narrows an outer union reference by a **nested** discriminant access, accepting code that `tsc` rejects. TypeScript narrows a union only by a discriminant property accessed **directly** on the reference (`x.kind === "a"`); it does **not** narrow the outer union through a nested access such as `x.meta.kind === "a"` — there the only references narrowed are `x.meta` and `x.meta.kind`, never `x`.

This is a parity false-negative: tsz silently accepts programs `tsc` flags with TS2339 / TS2322.

## Structural rule

When a condition tests `x.<p1>.<p2>…<pn> === <literal>` (or the `typeof`, aliased-const, or `switch` forms), `tsc` narrows the reference `x.<p1>…<p(n-1)>` (the **direct** parent of the discriminant property) by the single property `<pn>`. It never narrows an outer reference by a multi-segment discriminant path. tsz must narrow a union target by a discriminant only when the property path relative to that target has length 1.

## Repro (tsc 6.0.2, `--strict`)

```ts
type Variant = { meta: { kind: "a" }; a: number } | { meta: { kind: "b" }; b: string };
function visit(node: Variant): number | string {
  if (node.meta.kind === "a") return node.a; // tsc: TS2339 ; tsz: accepted (wrong)
  return node.b;                             // tsc: TS2339 ; tsz: accepted (wrong)
}
```

`tsc` output:
```
error TS2339: Property 'a' does not exist on type 'Variant'.
error TS2339: Property 'b' does not exist on type 'Variant'.
```

tsz (before fix): clean — it incorrectly narrows `node` to a single member via the nested `node.meta.kind` discriminant.

## Affected forms (all reproduce; `tsc` rejects, tsz accepted)

- Equality: `if (s.meta.kind === "a") …`
- Deeper nesting: `if (s.outer.inner.kind === "a") …`
- Aliased value: `const d = s.meta.kind; if (d === "a") …`
- Aliased object: `const m = s.meta; if (m.kind === "a") s.a` (narrows `m`, not `s`)
- `switch (s.meta.phase) { … }`
- `typeof s.payload.value === "string"`
- Member assignment: narrowed member is wrongly assignable (TS2322 in `tsc`).

Legitimate direct-discriminant narrowing is unaffected (`s.kind === "a"`, unit-typed `s.meta === "a"`, aliased top-level `const t = s.tag`, and narrowing the inner reference `s.meta` by its own direct discriminant all stay correct).

## Owner layer

Solver-backed flow analysis in `crates/tsz-checker/src/flow/control_flow/`. The discriminant-narrowing path producers walk the full property chain and apply a multi-segment path to the outer target; they must require a single-segment path.

## Acceptance

- The repro above emits TS2339 matching `tsc`.
- Direct/unit/aliased-top-level/inner-reference discriminant narrowing remain clean.
- Regression coverage with varied binder names plus positive controls.

--- a/crates/tsz-checker/Cargo.toml
+++ b/crates/tsz-checker/Cargo.toml
@@ -31,6 +31,10 @@ web-time = { workspace = true }
 stacker = "0.1.23"
 
 [[test]]
+name = "nested_discriminant_narrowing_tests"
+path = "tests/nested_discriminant_narrowing_tests.rs"
+
+[[test]]
 name = "inference_placeholder_determinism_tests"
 path = "tests/inference_placeholder_determinism_tests.rs"
 

--- a/crates/tsz-checker/src/flow/control_flow/narrowing.rs
+++ b/crates/tsz-checker/src/flow/control_flow/narrowing.rs
@@ -1350,14 +1350,23 @@ impl<'a> FlowAnalyzer<'a> {
             current = effective_target;
         }
 
-        // Reverse the path to get correct order (["payload", "kind"] not ["kind", "payload"])
-        path.reverse();
-
-        if path.is_empty() {
+        // Discriminant narrowing only applies to a *direct* property of the
+        // narrowed reference. `tsc` narrows a union by `x.kind === "a"` (a
+        // discriminant property accessed directly on `x`), but it does NOT
+        // narrow the outer union through a nested access such as
+        // `x.meta.kind === "a"` — there the only references narrowed are
+        // `x.meta` and `x.meta.kind`, never `x`. Requiring a single-segment
+        // path keeps every consumer (truthiness, switch, assertion predicate)
+        // aligned with that rule; nested references are narrowed when they are
+        // themselves the target via the relative-path producers.
+        //
+        // A single collected segment needs no reversal, so reject multi-segment
+        // paths before paying for `Vec::reverse`.
+        if path.len() != 1 {
             return None;
         }
 
-        // current is now the base (e.g., "action" in action.payload.kind)
+        // current is now the base (e.g., "action" in action.kind)
         Some((path, is_optional, current))
     }
 
@@ -1374,16 +1383,18 @@ impl<'a> FlowAnalyzer<'a> {
         // (discriminant_property_info returns the full path from the root, which is wrong when
         //  target is not the root — e.g., returns path=["test","type"] base=this for `this.test.type`
         //  when we need path=["type"] base=this.test relative to target=this.test)
+        // Single-segment relative path only (see `discriminant_property_info`):
+        // a nested access narrows the inner reference, never the outer `target`.
         if let Some(literal) = self.discriminant_literal_candidate(right)
             && let Some((rel_path, is_optional)) = self.relative_discriminant_path(left, target)
-            && !rel_path.is_empty()
+            && rel_path.len() == 1
         {
             return Some((rel_path, literal, is_optional, target));
         }
 
         if let Some(literal) = self.discriminant_literal_candidate(left)
             && let Some((rel_path, is_optional)) = self.relative_discriminant_path(right, target)
-            && !rel_path.is_empty()
+            && rel_path.len() == 1
         {
             return Some((rel_path, literal, is_optional, target));
         }
@@ -1482,8 +1493,10 @@ impl<'a> FlowAnalyzer<'a> {
                 || init_node.kind == syntax_kind_ext::ELEMENT_ACCESS_EXPRESSION
             {
                 // Walk from `init_expr` towards the root, collecting segments until we hit `target`.
+                // Single-segment path only: `const k = s.meta.kind` does not narrow `s` in `tsc`.
                 if let Some((rel_path, is_optional)) =
                     self.relative_discriminant_path(init_expr, target)
+                    && rel_path.len() == 1
                 {
                     return Some((rel_path, literal, is_optional, target));
                 }
@@ -1594,10 +1607,12 @@ impl<'a> FlowAnalyzer<'a> {
         right: NodeIndex,
         target: NodeIndex,
     ) -> Option<(Vec<Atom>, bool, &str)> {
+        // Single-segment path only (see `discriminant_property_info`):
+        // `typeof s.meta.x === "string"` narrows `s.meta`, never the outer union `s`.
         // Try left = typeof expr, right = string literal
         if let Some(operand) = self.get_typeof_operand(self.skip_parenthesized(left))
             && let Some((path, is_optional)) = self.relative_discriminant_path(operand, target)
-            && !path.is_empty()
+            && path.len() == 1
             && let Some(lit) = self.literal_string_from_node(right)
         {
             return Some((path, is_optional, lit));
@@ -1605,7 +1620,7 @@ impl<'a> FlowAnalyzer<'a> {
         // Try right = typeof expr, left = string literal
         if let Some(operand) = self.get_typeof_operand(self.skip_parenthesized(right))
             && let Some((path, is_optional)) = self.relative_discriminant_path(operand, target)
-            && !path.is_empty()
+            && path.len() == 1
             && let Some(lit) = self.literal_string_from_node(left)
         {
             return Some((path, is_optional, lit));

--- a/crates/tsz-checker/tests/nested_discriminant_narrowing_tests.rs
+++ b/crates/tsz-checker/tests/nested_discriminant_narrowing_tests.rs
@@ -1,0 +1,265 @@
+//! Regression coverage for discriminated-union narrowing through *nested*
+//! property accesses.
+//!
+//! Structural rule (matches TypeScript 6.0.x): a union reference is narrowed
+//! only by a discriminant property accessed **directly** on that reference
+//! (`x.kind === "a"`). TypeScript does NOT narrow an outer union through a
+//! nested discriminant access — `x.meta.kind === "a"` narrows `x.meta` (and
+//! `x.meta.kind`) but never `x`. Previously tsz walked the full property path
+//! (`["meta", "kind"]`) and narrowed the outer reference, accepting code that
+//! `tsc` rejects with TS2339/TS2322.
+//!
+//! Binder names are varied per case so the rule is structural, not keyed to a
+//! particular identifier.
+
+use tsz_checker::test_utils::check_source_strict_codes;
+
+const TS2339: u32 = 2339; // Property does not exist on type
+const TS2322: u32 = 2322; // Type not assignable
+
+fn codes(source: &str) -> Vec<u32> {
+    check_source_strict_codes(source)
+}
+
+// ---------------------------------------------------------------------------
+// Positive cases: legitimate direct-discriminant narrowing must keep working.
+// ---------------------------------------------------------------------------
+
+#[test]
+fn top_level_discriminant_still_narrows() {
+    // `tsc`: clean.
+    let diags = codes(
+        r#"
+type Shape = { kind: "circle"; radius: number } | { kind: "square"; side: string };
+function area(shape: Shape): number | string {
+    if (shape.kind === "circle") return shape.radius;
+    return shape.side;
+}
+"#,
+    );
+    assert!(
+        !diags.contains(&TS2339),
+        "direct discriminant `shape.kind` must still narrow `shape`, got {diags:?}"
+    );
+}
+
+#[test]
+fn unit_typed_intermediate_property_is_a_direct_discriminant() {
+    // `meta` itself is the unit discriminant (`"on" | "off"`), a direct property.
+    // `tsc`: clean.
+    let diags = codes(
+        r#"
+type Toggle = { meta: "on"; power: number } | { meta: "off"; reason: string };
+function read(node: Toggle): number | string {
+    if (node.meta === "on") return node.power;
+    return node.reason;
+}
+"#,
+    );
+    assert!(
+        !diags.contains(&TS2339),
+        "unit-typed direct property `node.meta` must narrow `node`, got {diags:?}"
+    );
+}
+
+#[test]
+fn narrowing_the_inner_reference_still_works() {
+    // The discriminant is direct *relative to the inner reference* `holder.inner`,
+    // so `tsc` narrows `holder.inner` and both branches type-check.
+    let diags = codes(
+        r#"
+type Inner = { kind: "a"; av: number } | { kind: "b"; bv: string };
+function pick(holder: { inner: Inner }): number | string {
+    if (holder.inner.kind === "a") return holder.inner.av;
+    return holder.inner.bv;
+}
+"#,
+    );
+    assert!(
+        !diags.contains(&TS2339),
+        "inner reference `holder.inner` must still narrow by its direct discriminant, got {diags:?}"
+    );
+}
+
+#[test]
+fn aliased_top_level_discriminant_still_narrows() {
+    // Aliased *direct* discriminant (TS 4.4 feature). `tsc`: clean.
+    let diags = codes(
+        r#"
+type Signal = { tag: "click"; x: number } | { tag: "key"; code: string };
+function handle(ev: Signal): number | string {
+    const t = ev.tag;
+    if (t === "click") return ev.x;
+    return ev.code;
+}
+"#,
+    );
+    assert!(
+        !diags.contains(&TS2339),
+        "aliased direct discriminant `const t = ev.tag` must narrow `ev`, got {diags:?}"
+    );
+}
+
+#[test]
+fn switch_on_direct_discriminant_still_narrows() {
+    // `tsc`: clean.
+    let diags = codes(
+        r#"
+type Cmd = { op: "add"; lhs: number } | { op: "neg"; value: string };
+function run(cmd: Cmd): number | string {
+    switch (cmd.op) {
+        case "add": return cmd.lhs;
+        case "neg": return cmd.value;
+    }
+}
+"#,
+    );
+    assert!(
+        !diags.contains(&TS2339),
+        "switch on direct discriminant `cmd.op` must narrow `cmd`, got {diags:?}"
+    );
+}
+
+// ---------------------------------------------------------------------------
+// Negative cases: nested discriminants must NOT narrow the outer union.
+// Each mirrors a `tsc` TS2339/TS2322 result.
+// ---------------------------------------------------------------------------
+
+#[test]
+fn nested_discriminant_does_not_narrow_outer_union() {
+    // `tsc`: TS2339 on `node.a` and `node.b` (no narrowing of `node`).
+    let diags = codes(
+        r#"
+type Variant = { meta: { kind: "a" }; a: number } | { meta: { kind: "b" }; b: string };
+function visit(node: Variant): number | string {
+    if (node.meta.kind === "a") return node.a;
+    return node.b;
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2339),
+        "nested discriminant `node.meta.kind` must NOT narrow `node`; expected TS2339, got {diags:?}"
+    );
+}
+
+#[test]
+fn deeply_nested_discriminant_does_not_narrow_outer_union() {
+    // `tsc`: TS2339 (two levels of nesting).
+    let diags = codes(
+        r#"
+type Wrap =
+    | { outer: { inner: { kind: "a" } }; a: number }
+    | { outer: { inner: { kind: "b" } }; b: string };
+function take(w: Wrap): number | string {
+    if (w.outer.inner.kind === "a") return w.a;
+    return w.b;
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2339),
+        "deep nested discriminant must NOT narrow the root; expected TS2339, got {diags:?}"
+    );
+}
+
+#[test]
+fn aliased_nested_discriminant_does_not_narrow_outer_union() {
+    // `const disc = action.meta.type` does not narrow `action`. `tsc`: TS2339.
+    let diags = codes(
+        r#"
+type Action = { meta: { type: "inc" }; amount: number } | { meta: { type: "dec" }; id: string };
+function reduce(action: Action): number | string {
+    const disc = action.meta.type;
+    if (disc === "inc") return action.amount;
+    return action.id;
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2339),
+        "aliased nested discriminant must NOT narrow the root; expected TS2339, got {diags:?}"
+    );
+}
+
+#[test]
+fn aliased_intermediate_object_does_not_narrow_outer_union() {
+    // `const m = item.meta; if (m.kind === ...)` narrows `m`, not `item`. `tsc`: TS2339.
+    let diags = codes(
+        r#"
+type Item = { meta: { kind: "a" }; a: number } | { meta: { kind: "b" }; b: string };
+function use(item: Item): number | string {
+    const m = item.meta;
+    if (m.kind === "a") return item.a;
+    return item.b;
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2339),
+        "aliasing the intermediate object must NOT narrow the root; expected TS2339, got {diags:?}"
+    );
+}
+
+#[test]
+fn switch_on_nested_discriminant_does_not_narrow_outer_union() {
+    // `tsc`: TS2339 — switching on `state.meta.phase` does not narrow `state`.
+    let diags = codes(
+        r#"
+type State = { meta: { phase: "load" }; progress: number } | { meta: { phase: "done" }; result: string };
+function summarize(state: State): number | string {
+    switch (state.meta.phase) {
+        case "load": return state.progress;
+        case "done": return state.result;
+    }
+    return 0;
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2339),
+        "switch on nested discriminant must NOT narrow the root; expected TS2339, got {diags:?}"
+    );
+}
+
+#[test]
+fn nested_discriminant_does_not_assign_narrowed_member() {
+    // The narrowed member is not assignable from the un-narrowed union. `tsc`: TS2322.
+    let diags = codes(
+        r#"
+type Add = { meta: { kind: "a" }; a: number };
+type Op = Add | { meta: { kind: "b" }; b: string };
+function go(op: Op): void {
+    if (op.meta.kind === "a") {
+        const only: Add = op;
+        void only;
+    }
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2322),
+        "outer union must stay un-narrowed through a nested discriminant; expected TS2322, got {diags:?}"
+    );
+}
+
+#[test]
+fn typeof_nested_property_does_not_narrow_outer_union() {
+    // `typeof box.payload.value === "string"` narrows `box.payload`, never `box`.
+    // `tsc`: TS2339 on `box.s` / `box.n`.
+    let diags = codes(
+        r#"
+type Box =
+    | { payload: { value: string }; s: number }
+    | { payload: { value: number }; n: string };
+function unwrap(box: Box): number | string {
+    if (typeof box.payload.value === "string") return box.s;
+    return box.n;
+}
+"#,
+    );
+    assert!(
+        diags.contains(&TS2339),
+        "typeof on a nested property must NOT narrow the root; expected TS2339, got {diags:?}"
+    );
+}

